Fixed #5621 - toLocaleString now accesses object property

Dachande663 opened the pull request:
As per issue https://github.com/Microsoft/ChakraCore/issues/5621, this PR enables toLocaleString to call JavascriptString underlying property.

sethbrenith commented:
Thanks @Dachande663 . Would you mind also adding unit tests as part of this change? It looks like test/Array/toLocaleString.js would be an appropriate place. The test file basically just prints out text and the unit test runner compares the output text against the expected text in test/Array/toLocaleString.baseline. We should test at least these cases (feel free to add other relevant cases if you think of them):

* The case from #5621
* If we set `String.prototype.toLocaleString = undefined`, then `["test"].toLocaleString()` should throw a TypeError

sethbrenith commented on lib/Runtime/Language/JavascriptConversion.cpp:
```diff
@@ -701,8 +701,16 @@ using namespace Js;
 
-        case TypeIds_String:
-            return UnsafeVarTo<JavascriptString>(aValue);
+        case TypeIds_String: {
```
nit: we put opening `{` on the following line

sethbrenith replied:
This seems very similar to the `default` case below. Why not just remove the `TypeIds_String` case?

sethbrenith commented on lib/Runtime/Language/JavascriptConversion.cpp:
```diff
@@ -701,8 +701,16 @@ using namespace Js;
+        case TypeIds_String: {
+            RecyclableObject* object = VarTo<RecyclableObject>(aValue);
+            Var value = JavascriptOperators::GetProperty(object, PropertyIds::toLocaleString, scriptContext, NULL);
+            RecyclableObject* toLocaleStringFunction = VarTo<RecyclableObject>(value);
```
`VarTo<RecyclableObject>(value)` will crash the program if `value` is not a recyclable object (suppose maybe it's an int, for example).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/ChakraCore?pullRequest=5742) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/ChakraCore?pullRequest=5742) before we can accept your contribution.<br/><br/>:x: Dachande663 [sign now](https://cla.opensource.microsoft.com/Microsoft/ChakraCore?pullRequest=5742)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/ChakraCore?pullRequest=5742) it.</sub>

Dachande663 commented:
@sethbrenith thanks for the feedback. I've removed the String case allowing it to fallback to default and also added in tests. I'm not sure what isolation the test runner uses in between scenarios so let me know if overloading the String prototype will have any adverse effects.

sethbrenith commented on test/Array/toLocaleString.baseline:
```diff
@@ -43,3 +43,15 @@ TypeError : The value of the property 'toLocaleString' is not a Function object
+
+
+11. Array: toLocaleString should throw TypeError on undefined
+TypeError : Cannot redefine non-configurable property 'toLocaleString'
```
This is a TypeError but not the one I was hoping to trigger. This is happening in the `Object.defineProperty` step, because the previous scenario used `Object.defineProperty` also and didn't specify `configurable: true`, meaning the property can never be reconfigured again.

sethbrenith commented:
The test runner runs each .js file in an entirely new process, so there's no risk of prototype modifications leaking to other tests. However, the scenarios within this test file don't have any enforced isolation; they're just to make the output look pretty. Ideally we should have each test scenario put things back the way they were, so that if somebody decides to reorder tests or add new ones on the end they wouldn't have any nasty surprises. Something like:

```javascript
var originalToLocaleString = String.prototype.toLocaleString;
guarded_call(function () {
    // test contents that might change prototypes...
});
String.prototype.toLocaleString = originalToLocaleString;
```

sethbrenith commented:
/cc @jackhorton

Dachande663 commented:
@sethbrenith roger. I've updated the test suite to actually do the tests needed and ensure state is returned afterwards. Cheers for taking the time to provide the comments, massively helpful.

sethbrenith commented:
@dotnet-bot test Windows 10 x64_test please
